src/components/form-control/form-control.jsx

```diff
@@ -1,80 +1,63 @@
-}
+const FormControl = forwardRef(function FormControl(props, ref) {
+  const classNames = {
+    container: props.className,
+    control: styles['control-container'],
+    heading: styles['label-wrapper'],
+    label: props.error ? styles['invalid-label'] : styles.label,
+  };
+  const ids = {
+    input: props.id,
+    label: props.labelId || `${props.id}-label`,
+    tooltip: `${props.id}-tooltip`,
+  };
 
-function isInvalid(error, readOnly) {
-  return !!error && !readOnly;
-}
+  useImperativeHandle(ref, () => ({
+    /** Determine whether the component has a different value than the provided prop. */
+    get dirty() { return undefined; },
+    /** The ID of the input element. */
+    id: props.id, // Is this needed anymore? Seems like it can be solved by `name`
+    /** The name of the control element which is used to reference the form-data after submitting the form. */
+    name: props.name,
+    /** Access the value of the component without hoisting value state */
+    get value() { return undefined; },
+  }));
 
-export default function FormControl(props) {
   return (
-    <div className={props.className} onKeyDown={props.onKeyDown} role="presentation">
-      <div className={styles['label-wrapper']}>
-        <div className={getLabelClassName(props.error, props.readOnly)}>
-          <label
-            htmlFor={props.id}
-            id={props.labelId}
-          >
+    <div className={classNames.container} onKeyDown={props.onKeyDown} role="presentation">
+      <div className={classNames.heading}>
+        <div className={classNames.label}>
+          <label htmlFor={ids.input} id={ids.label}>
             {props.label}
           </label>
           {props.required && '(Required)'}
         </div>
         {!!props.tooltip && (
-          <HelpIcon id={`${props.id}-tooltip`} label="More information" text={props.tooltip} />
+          <HelpIcon id={ids.tooltip} label="More information" text={props.tooltip} />
         )}
       </div>
-      <div className={styles['control-container']}>
+      <div className={classNames.control}>
         {props.children}
       </div>
-      {isInvalid(props.error, props.readOnly) && (
-        <span
-          id={`${props.id}Hint`}
-          className={styles['error-message']}
-          aria-live="polite"
-        >
-          {props.error}
-        </span>
-      )}
     </div>
   );
-}
+});
 
 FormControl.propTypes = {
   className: PropTypes.string,
   children: PropTypes.node.isRequired,
-  error: PropTypes.string,
-  id: (props) => {
-    if (props.id === undefined && props.labelId === undefined) {
-      return new Error('Invalid prop `id` supplied to `FormControl`. Either `id` or `labelId` are required.');
-    }
-
-    if (typeof props.id !== 'string' && props.labelId === undefined) {
-      return new Error('Invalid prop `id` supplied to `FormControl`. `id` must be a string.');
-    }
-
-    return undefined;
-  },
+  error: PropTypes.string, // TODO: Refactor as boolean
```

Comment:
Correct! `FormControl` now only cares about "truthy" value instead of the actual error message. But I might regret this if it makes it harder to grok.